src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs

```diff
@@ -1440,41 +1511,30 @@ private static unsafe nuint NarrowUtf16ToAscii_Sse2(char* pUtf16Buffer, byte* pA
             // jumps as much as possible in the optimistic case of "all ASCII". If we see non-ASCII
             // data, we jump out of the hot paths to targets at the end of the method.
 
-            Debug.Assert(Sse2.IsSupported);
-            Debug.Assert(BitConverter.IsLittleEndian);
+            Debug.Assert(Sse2.IsSupported || AdvSimd.Arm64.IsSupported, "Sse2 or AdvSimd64 required.");
+            Debug.Assert(BitConverter.IsLittleEndian, "This SSE2/Arm64 implementation assumes little-endian.");
             Debug.Assert(elementCount >= 2 * SizeOfVector128);
 
             Vector128<short> asciiMaskForTestZ = Vector128.Create(unchecked((short)0xFF80)); // used for PTEST on supported hardware
             Vector128<ushort> asciiMaskForAddSaturate = Vector128.Create((ushort)0x7F80); // used for PADDUSW
-            const int NonAsciiDataSeenMask = 0b_1010_1010_1010_1010; // used for determining whether the pmovmskb operation saw non-ASCII chars
 
             // First, perform an unaligned read of the first part of the input buffer.
-
-            Vector128<short> utf16VectorFirst = Sse2.LoadVector128((short*)pUtf16Buffer); // unaligned load
+            ref ushort utf16Buffer = ref *(ushort*)pUtf16Buffer;
+            ref byte asciiBuffer = ref *pAsciiBuffer;
+            Vector128<ushort> utf16VectorFirst = Vector128.LoadUnsafe(ref utf16Buffer); // unaligned load
 
             // If there's non-ASCII data in the first 8 elements of the vector, there's nothing we can do.
             // See comments in GetIndexOfFirstNonAsciiChar_Sse2 for information about how this works.
 
-            if (Sse41.IsSupported)
+            if (VectorContainsNonAsciiChar(utf16VectorFirst, asciiMaskForTestZ, asciiMaskForAddSaturate))
             {
-                if (!Sse41.TestZ(utf16VectorFirst, asciiMaskForTestZ))
-                {
-                    return 0;
-                }
-            }
-            else
-            {
-                if ((Sse2.MoveMask(Sse2.AddSaturate(utf16VectorFirst.AsUInt16(), asciiMaskForAddSaturate).AsByte()) & NonAsciiDataSeenMask) != 0)
-                {
-                    return 0;
-                }
+                return 0;
             }
 
             // Turn the 8 ASCII chars we just read into 8 ASCII bytes, then copy it to the destination.
 
-            Vector128<byte> asciiVector = Sse2.PackUnsignedSaturate(utf16VectorFirst, utf16VectorFirst);
-            Sse2.StoreScalar((ulong*)pAsciiBuffer, asciiVector.AsUInt64()); // ulong* calculated here is UNALIGNED
-
+            Vector128<byte> asciiVector = ExtractAsciiVector(utf16VectorFirst);
+            asciiVector.GetLower().StoreUnsafe(ref asciiBuffer);
```

Comment:
👍. `Vector64<T>` is a bit odd on x86/x64 since its treated as a regular `struct { ulong _value; }`, so I'd expect promotion/etc to do the right thing here
but it might not be as efficient as extracting the lowest `UInt64` scalar (it probably could be with some tweaks however if it does differ)

Reply:
Seems that by switching to `StoreUnsafe` introduced a bigger sequence of assembly with a few unnecessary movs.
On HEAD, we get the following assembly for extract and store.
```
   vpackuswb xmm0, xmm0, xmm0
    vmovq    qword ptr [r15], xmm0
    mov      r12d, 8
    test     r15b, 8
```

With the above change, extract and store looks as following.
```
mov      r12, rbx
    vpackuswb xmm0, xmm0, xmm0
    vmovapd  xmmword ptr [rbp-80H], xmm0
    mov      rax, qword ptr [rbp-80H]
    mov      qword ptr [rbp-58H], rax
    mov      rax, qword ptr [rbp-58H]
    mov      qword ptr [r12], rax
    mov      r13d, 8
    test     bl, 8
```
Would you suggest to stick to Vector128 api in this case? 
Full assembly dumps are available here - [HEAD](https://gist.github.com/SwapnilGaikwad/1d294bdcac31b91d2a83266f09aef061), [PR](https://gist.github.com/SwapnilGaikwad/a16089e23cd9d57145c0afb591ad7a7f)

Reply:
I meant - can you share the arm64 versions? You can also turn off the address displays using `set COMPlus_JitDiffableDasm=1`.

Reply:
Assembly dumps without `Debug.Assert()`- [HEAD](https://gist.github.com/SwapnilGaikwad/6bd218c5113f622e5aafce6bb1feca7b), [PR](https://gist.github.com/SwapnilGaikwad/e4e65798b924567c57cd0588a5c4740a).
Having assert introduced additional [stack mov](https://gist.github.com/SwapnilGaikwad/a16089e23cd9d57145c0afb591ad7a7f#file-less_args_asm_dump-txt-L119). Commenting debug asserts may resemble closely to the assembly with the release builds.

Reply:
> Would you suggest to stick to Vector128 api in this case?

That's probably the best for xplat right now. Ideally we'd also handle `Vector64<T>` more optimally on x86/x64, but that's not likely for .NET 7

Reply:
> That's probably the best for xplat right now. Ideally we'd also handle `Vector64<T>` more optimally on x86/x64, but that's not likely for .NET 7

Sure. However, additional instructions with the new type of store might cause a bit slowdown. Will that be alright? I couldn't get the consistent numbers between the performance runs but saw slowdown between 0% to 4%.
Results from the 3 executions.
- less-args-vectorContains (PR)
- main (HEAD)

```
|   Method |        Job |                                                                                                  Toolchain | size | encName |     Mean |    Error |   StdDev |   Median |      Min |      Max | Ratio | MannWhitney(2%) | RatioSD |  Gen 0 | Allocated | Alloc Ratio |
|--------- |----------- |----------------------------------------------------------------------------------------------------------- |----- |-------- |---------:|---------:|---------:|---------:|---------:|---------:|------:|---------------- |--------:|-------:|----------:|------------:|
| GetBytes | Job-HXUYDD | /less-args-vectorContains/bin/testhost/net7.0-Linux-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun |   16 |   ascii | 25.33 ns | 0.221 ns | 0.206 ns | 25.34 ns | 24.94 ns | 25.66 ns |  1.02 |            Same |    0.02 | 0.0039 |      40 B |        1.00 |
| GetBytes | Job-ISMZPV |                     /main/bin/testhost/net7.0-Linux-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun |   16 |   ascii | 24.83 ns | 0.346 ns | 0.323 ns | 24.81 ns | 24.45 ns | 25.66 ns |  1.00 |            Base |    0.00 | 0.0039 |      40 B |        1.00 |
|          |            |                                                                                                            |      |         |          |          |          |          |          |          |       |                 |         |        |           |             |
| GetBytes | Job-HXUYDD | /less-args-vectorContains/bin/testhost/net7.0-Linux-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun |  512 |   ascii | 90.16 ns | 1.140 ns | 1.066 ns | 89.95 ns | 89.01 ns | 92.63 ns |  1.02 |            Same |    0.01 | 0.0532 |     536 B |        1.00 |
| GetBytes | Job-ISMZPV |                     /main/bin/testhost/net7.0-Linux-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun |  512 |   ascii | 88.62 ns | 0.428 ns | 0.380 ns | 88.65 ns | 87.92 ns | 89.12 ns |  1.00 |            Base |    0.00 | 0.0530 |     536 B |        1.00 |

|   Method |        Job |                                                                                                  Toolchain | size | encName |     Mean |    Error |   StdDev |   Median |      Min |      Max | Ratio | MannWhitney(2%) |  Gen 0 | Allocated | Alloc Ratio |
|--------- |----------- |----------------------------------------------------------------------------------------------------------- |----- |-------- |---------:|---------:|---------:|---------:|---------:|---------:|------:|---------------- |-------:|----------:|------------:|
| GetBytes | Job-ZNCMEJ | /less-args-vectorContains/bin/testhost/net7.0-Linux-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun |   16 |   ascii | 25.08 ns | 0.355 ns | 0.332 ns | 24.91 ns | 24.69 ns | 25.82 ns |  1.00 |            Same | 0.0039 |      40 B |        1.00 |
| GetBytes | Job-MYVQNW |                     /main/bin/testhost/net7.0-Linux-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun |   16 |   ascii | 25.16 ns | 0.258 ns | 0.242 ns | 25.10 ns | 24.88 ns | 25.63 ns |  1.00 |            Base | 0.0040 |      40 B |        1.00 |
|          |            |                                                                                                            |      |         |          |          |          |          |          |          |       |                 |        |           |             |
| GetBytes | Job-ZNCMEJ | /less-args-vectorContains/bin/testhost/net7.0-Linux-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun |  512 |   ascii | 90.92 ns | 0.481 ns | 0.450 ns | 90.97 ns | 90.22 ns | 91.76 ns |  1.00 |            Same | 0.0529 |     536 B |        1.00 |
| GetBytes | Job-MYVQNW |                     /main/bin/testhost/net7.0-Linux-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun |  512 |   ascii | 90.89 ns | 0.751 ns | 0.702 ns | 90.53 ns | 90.06 ns | 92.11 ns |  1.00 |            Base | 0.0529 |     536 B |        1.00 |

|   Method |        Job |                                                                                                  Toolchain | size | encName |     Mean |    Error |   StdDev |   Median |      Min |       Max | Ratio | MannWhitney(2%) | RatioSD |  Gen 0 | Allocated | Alloc Ratio |
|--------- |----------- |----------------------------------------------------------------------------------------------------------- |----- |-------- |---------:|---------:|---------:|---------:|---------:|----------:|------:|---------------- |--------:|-------:|----------:|------------:|
| GetBytes | Job-RXUBVW | /less-args-vectorContains/bin/testhost/net7.0-Linux-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun |   16 |   ascii | 25.03 ns | 0.274 ns | 0.243 ns | 25.01 ns | 24.72 ns |  25.52 ns |  1.02 |            Same |    0.01 | 0.0040 |      40 B |        1.00 |
| GetBytes | Job-SVUNDM |                     /main/bin/testhost/net7.0-Linux-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun |   16 |   ascii | 24.59 ns | 0.219 ns | 0.194 ns | 24.58 ns | 24.28 ns |  24.94 ns |  1.00 |            Base |    0.00 | 0.0039 |      40 B |        1.00 |
|          |            |                                                                                                            |      |         |          |          |          |          |          |           |       |                 |         |        |           |             |
| GetBytes | Job-RXUBVW | /less-args-vectorContains/bin/testhost/net7.0-Linux-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun |  512 |   ascii | 89.57 ns | 0.928 ns | 0.868 ns | 89.45 ns | 88.64 ns |  91.22 ns |  1.04 |          Slower |    0.01 | 0.0531 |     536 B |        1.00 |
| GetBytes | Job-SVUNDM |                     /main/bin/testhost/net7.0-Linux-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun |  512 |   ascii | 86.08 ns | 0.753 ns | 0.705 ns | 86.05 ns | 84.91 ns |  87.50 ns |  1.00 |            Base |    0.00 | 0.0530 |     536 B |        1.00 |


```

Reply:
What code are you using for extract and store?

Reply:
> used the release builds for performance runs while created separate checked builds to compare the assembly. I am not sure which libclrjit.so needs to be replaced with the checked version.

Well, you can just use release binaries (the way you did for performance runs), but just replace the release version of `libclrjit.so` with checked version of `libclrjit.so` so you can use `COMPlus_JitDisasm` flag. That flag is only for checked or debug build and not in release build.

Reply:
cool, let me try it.

Reply:
> What code are you using for extract and store?

Head does the following
```
Vector128<byte> asciiVector = Sse2.PackUnsignedSaturate(utf16VectorFirst, utf16VectorFirst);
Sse2.StoreScalar((ulong*)pAsciiBuffer, asciiVector.AsUInt64()); // ulong* calculated here is UNALIGNED
```
and the patch does this
```
ref byte asciiBuffer = ref *pAsciiBuffer;
Vector128<byte> asciiVector = ExtractAsciiVector(utf16VectorFirst);
 asciiVector.GetLower().StoreUnsafe(ref asciiBuffer);
```
[Link to diff](https://github.com/dotnet/runtime/pull/70080/files#diff-66bbe89271f826c9232bd146abb678844754515dc027f70ad0ce36f751da46ebL1475-R1478)

Reply:
[Here](https://gist.github.com/SwapnilGaikwad/f5533c5e57e0a277ef26305e716c730d) is the updated one. For the PR, assembly for `NarrowUtf16ToAscii_Intrinsified` should have been dumped instead of `NarrowUtf16ToAscii`.

Reply:
I believe this one uses `StoreUnsafe` because I don't see AV related code generated as you pointed out in https://github.com/dotnet/runtime/pull/70080#discussion_r906229412?

Reply:
Yup, this one uses `StoreUnsafe`. The aligned stores are only performed in a loop, now changing them to `StoreUnsafe`.